Add ABI language to the standard (#375)
pmix/pmix-standard: Chap_Introduction.tex

To facilitate such interoperability the PMIx \ac{ASC} maintains a set of standardized headers that are versioned with the PMIx Standard that applications and tools can reference~\footnote{PMIx Headers for ABI Compatibility \url{https://github.com/pmix/pmix-abi}}.

In recognition that there are circumstances where the \ac{ABI} needs to be modified this section defines some guidance for making such modifications.
Additions to the PMIx interface can occur without breaking \ac{ABI} compatibility.
Deprecating and removing portions of the PMIx interface does break \ac{ABI} compatibility.
Modifications to the existing PMIx interface do break \ac{ABI} compatibility.
Any changes to the \ac{ABI} will occur in a major release and will be noted in the release notes.

The PMIx \ac{ABI} is comprised of the following:
\begin{compactitemize}
\item Function signatures
\item Constants and their values
\item Structures and their membership (including field position)
\item Attributes that are required to be implemented and their string representation
\end{compactitemize}

> If a data structure must be changed in a way that breaks the \ac{ABI} the authors should consider creating a new data structure if possible.

Reviewer: Changing the ABI of data structures (e.g., removing/reordering existing fields) will break backward compatibility. We should add a short rationale as to why this is **very** undesirable, and not only one should *consider*, but should avoid this type of changes at all cost.

Author: ```suggestion
If a data structure must be changed in a way that breaks the \ac{ABI} the authors are strongly advised to create a new data structure instead of breaking the \ac{ABI} and backward compatibility.
```

Let's add a clear rationale to add to this section to capture the middleware constraints dealing with cross and backwards compatibility issues.

Reviewer: ```diff
+Changing data structures (in particular changes that modify the field alignment) and function signatures
+breaks the \ac{ABI}. The \ac{ASC} strongly discourages such changes. It is preferred that the authors 
+duplicate existing PMIx interfaces that are impacted by the change, 
+and potentially initiate the deprecation procedure for the duplicated interfaces.
+
+\rationalestart
+Breaking the \ac{ABI} of data structures and function signatures while using the same
+name prevents the implementation of backward compatible and cross-version compatible
+\ac{PMIx} libraries. 
+\rationaleend
```

Reviewer: I like the new text but we may need some more plenary time to discuss the implications.

Author: ```suggestion
The \ac{ASC} prohibits data structure and function signature changes that break the \ac{ABI}
(e.g., modifications to data structure member alignment and/or size).
It is preferred that the authors create a new data structure or function and
duplicate all impacted PMIx interfaces.

\rationalestart
Breaking the \ac{ABI} of data structures and function signatures while using the same
name prevents a PMIx implementation from supporting backward compatible and
cross-version compatible \ac{PMIx} libraries. 
\rationaleend
```

> Additionally, the authors must take care to update or duplicate existing PMIx interfaces that are impacted by the change.

Author: ```suggestion
```

> Attributes that are required by the PMIx standard are a required part of the \ac{ABI}.

Reviewer: Ralph comment: Attributes are non-linkable string objects (i.e., they are defines, they don't have an address as a variable), hence they are not an ABI element.

Reviewer: ```Although the values of attributes and constants are not part of the \ac{ABI}, they are defined as standardized members of the \ac{API}, hence their value can be assumed based on the \ac{ABI} version.```

Author: Please take time before Wednesday afternoon to suggest any revisions to this text.

This text will replace the full paragraph here (lines 266-271).

Author: I captured this change in the https://github.com/pmix/pmix-standard/pull/375/files#r870587051
Let's follow up on that comment.

> Attributes that are required by the PMIx standard are a required part of the \ac{ABI}.
> Attributes that are not required by the PMIx standard may be included in the \ac{ABI} while others may not.
> Attributes that are part of the \ac{ABI} are denoted as such in their description.
> PMIx attributes that are not required by the PMIx standard have a well-defined behavior (see Section~\ref{chap:intro:attributes}) when the string representation is not recognized by the implementation.
> The caller of the interface already expects to respond to the scenario in which the implementation does not support that optional attribute.
> Excluding these optional attributes from the ABI allows an implementation to provide additional attributes that are non-standard.

Author: ```suggestion
The value associated with attributes and constants are part of an \ac{API} and not part of a \refterm{Linker ABI}.
The PMIx Standard defines a \refterm{Build ABI} where the value associated with PMIx attributes and constants can be assumed based on the \ac{ABI} version.
```

The PMIx Standard \ac{ABI} is defined by a set of \ac{ASC} maintained headers.
The instantiation of the PMIx Standard \ac{ABI} in binary form is platform-specific.
The PMIx Standard and PMIx implementations must take care with data structures (e.g., \refstruct{pmix_value_t}) to ensure that the offset and size of members within the structure along with the size of the structure itself remain stable for a given platform.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection{PMIx ABI Versioning}
\label{chap:intro:abi:versioning}

The PMIx Standard ABI is defined in two parts.
The \textit{PMIx Standard Stable ABI} represents the Stable PMIx Standard elements (see the PMIx Governance document).
The \textit{PMIx Standard Provisional ABI} represents the Provisional PMIx Standard elements (see the PMIx Governance document).

Both the Stable ABI and Provisional ABI are versioned with two increasing numbers:
\begin{compactitemize}
\item \texttt{MAJOR} incremented when the ABI changed in a backward-incompatible manner.
\item \texttt{MINOR} incremented when functionality is added to the ABI in a backward-compatible manner.
\end{compactitemize}

> 

Author: Add a comment along the lines of:
```
The PMIx Standard ABI versioning follows the Semantic versioning procedure. However, the PMIx Implementation may use a different version numbering for their library objects than the PMix Standard ABI.
```

Author: Add the comment:
```suggestion
The PMIx Standard ABI version numbers largely follow the Semantic Versioning 2.0.0 specification\footnote{Semantic Versioning \url{https://semver.org/spec/v2.0.0.html}}.
However, a PMIx implementation may use a different version numbering technique for the objects associated with that implementation.
```

In recognition that there are circumstances where the \ac{ABI} needs to be modified this section defines some guidance for making such modifications.
Additions to the PMIx interface can occur without breaking \ac{ABI} compatibility.
Deprecating and removing portions of the PMIx interface does break \ac{ABI} compatibility.
Modifications to the existing PMIx interface does break \ac{ABI} compatibility.
Any changes to the \ac{ABI} will occur in a major release, and will be noted in the release notes.

The PMIx \ac{ABI} is comprised of the following:
\begin{compactalphaenum}
\item Function signatures
\item Constants and their values
\item Structures and their membership (including field position)
\item Attributes that are required to be implemented and their string representation
\end{compactalphaenum}

When adding a field to an existing data structure that has padding then that padding must be adjusted so as to preserve the stability of the \ac{ABI}.
If a data structure must be changed in a way that breaks the \ac{ABI} the authors should consider creating a new data structure if possible.

> Additionally, the authors must take care to update or duplicate existing PMIx interfaces that are impacted by the change.

Author: Remove. No padding in the current standard. May add in the future if padding in a structure is added.
